src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -257,29 +348,170 @@ public void WriteValue(byte[] value)
+                // this is an array, unless this is the first item in the array.
+                this.bufferWriter.Write(itemSeparator.Slice(0, 1).Span);
+            }
+
+            this.bufferWriter.Write(Encoding.UTF8.GetBytes(rawValue));
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            if (this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray || !this.IsInArray())
+            {
+                this.shouldWriteSeparator = true;
+            }
+
+            if (this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray)
+            {
+                this.isWritingConsecutiveRawValuesAtStartOfArray = true;
+            }
+
+            this.isWritingAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Manually writes an item separator ','
+        /// to the output if necessary, bypassing the Utf8JsonWriter.
+        /// Should be used before property names in an object or values
+        /// in an array.
+        /// This method should not be called by <see cref="WriteRawValue(string)"/>.
+        /// </summary>
+        private void WriteSeparatorIfNecessary()
+        {
+            if (this.shouldWriteSeparator)
+            {
+                this.WriteItemSeparator();
+            }
+
+            // reset state since now we're back in sync with Utf8JsonWriter.
+            this.shouldWriteSeparator = false;
+            this.isWritingAtStartOfArray = false;
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+        }
+
+        /// <summary>
+        /// Manually writes an item separator ',' to the output, bypassing the Utf8JsonWriter.
+        /// </summary>
+        private void WriteItemSeparator()
+        {
+            this.CommitUtf8JsonWriterContentsToBuffer();
+            Span<byte> buf = this.bufferWriter.GetSpan(1);
+            buf[0] = (byte)',';
+            this.bufferWriter.Advance(1);
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an object.
+        /// </summary>
+        private void EnterObjectScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingAtStartOfArray = false;
+            this.bitStack.PushTrue();
+        }
+
+        /// <summary>
+        /// Mark that the writer has transitioned to write an array.
+        /// </summary>
+        private void EnterArrayScope()
+        {
+            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
+            this.isWritingAtStartOfArray = true;
+            this.bitStack.PushFalse();
+        }
+
+        /// <summary>
+        /// Exit the current object or array scope.
+        /// </summary>
+        private void ExitScope()
+        {
+            this.isWritingAtStartOfArray = false;

```

Comment:
```suggestion
            this.isWritingAtStartOfArray = false;
            this.isWritingConsecutiveRawValuesAtStartOfArray = false;
```
For our piece of mind...

Reply:
Or maybe that could introduce a bug... Please double-check